Cannot adapt c array (stack)

The following code doesn't work

``` 
    double data[4] = {1, 2, 3, 4};
    std::vector<std::size_t> shape = { size };
    auto a = xt::adapt(data, size, xt::no_ownership(), shape);
    std::cout << a << std::endl;

xbuffer_adaptor.hpp:282:57: error: cannot convert ‘double*’ to ‘xt::detail::xbuffer_storage<double*&, std::allocator<double [4]> >::pointer {aka double (*)[4]}’ in initialization
             : p_data(std::forward<P>(data)), m_size(size)

```

However if I use heap allocated array (`double* data = new double[4];`)  it works.  

It seems I have to pass in `&data[0]` manually, maybe the template picks up a wrong function.

--- a/docs/source/adaptor.rst
+++ b/docs/source/adaptor.rst
@@ -144,3 +144,33 @@ adaptor before calling ``compute`` and pass it to the function:
         std::cout << std::endl;
         // prints 0 1 0 2
     }
+
+Adapting stack-allocated arrays
+-------------------------------
+
+Adapting C arrays allocated on the stack is as simple as adapting ``std::vector``:
+
+.. code::
+
+    #include <cstddef>
+    #include <vector>
+    #include "xtensor/xarray.hpp"
+    #include "xtensor/xadapt.hpp"
+
+    double v[6] = {1., 2., 3., 4., 5., 6. };
+    std::vector<std::size_t> shape = { 2, 3 };
+    auto a1 = xt::adapt(v, shape);
+
+    xt::xarray<double> a2 = {{ 1., 2., 3.},
+                             { 4., 5., 6.}};
+
+    xt::xarray<double> res = a1 + a2;
+    // res = {{ 2., 4., 6. }, { 8., 10., 12. }};
+
+``v`` is not copied into ``a1``, so if you change a value in ``a1``, you're actually changing
+the corresponding value in ``v``:
+
+.. code::
+
+    a1(0, 0) = 20.;
+    // now v is { 20., 2., 3., 4., 5., 6. }

--- a/docs/source/api/xarray_adaptor.rst
+++ b/docs/source/api/xarray_adaptor.rst
@@ -29,3 +29,9 @@ Defined in ``xtensor/xadapt.hpp``
 
 .. doxygenfunction:: xt::adapt(P&&, typename A::size_type, O, SC&&, SS&&, const A&)
    :project: xtensor
+
+.. doxygenfunction:: xt::adapt(T (&)[N], const SC&, layout_type)
+   :project: xtensor
+
+.. doxygenfunction:: xt::adapt(T (&)[N], SC&&, SS&&)
+   :project: xtensor

--- a/docs/source/api/xtensor_adaptor.rst
+++ b/docs/source/api/xtensor_adaptor.rst
@@ -35,3 +35,9 @@ Defined in ``xtensor/xadapt.hpp``
 
 .. doxygenfunction:: xt::adapt(P&&, typename A::size_type, O, SC&&, SS&&, const A&)
    :project: xtensor
+
+.. doxygenfunction:: xt::adapt(T (&)[N], const SC&, layout_type)
+   :project: xtensor
+
+.. doxygenfunction:: xt::adapt(T (&)[N], SC&&, SS&&)
+   :project: xtensor

--- a/docs/source/numpy-differences.rst
+++ b/docs/source/numpy-differences.rst
@@ -30,7 +30,7 @@ expressions can be accessed in the same way as values of higher dimensional arra
 Accumulators (``cumsum``, ``cumprod``) throw an exception if an axis argument is passed and the
 array argument is a 0-D argument:
 
-.. code-block::
+.. code::
 
     #include <xtensor/xarray.hpp>
     #include <xtensor/xio.hpp>

--- a/include/xtensor/xadapt.hpp
+++ b/include/xtensor/xadapt.hpp
@@ -138,7 +138,7 @@ namespace xt
 
     /**
      * Constructs an xarray_adaptor of the given dynamically allocated C array,
-     * with the specified shape and layout.
+     * with the specified shape and strides.
      * @param pointer the pointer to the beginning of the dynamic array
      * @param size the size of the dynamic array
      * @param ownership indicates whether the adaptor takes ownership of the array.
@@ -160,6 +160,37 @@ namespace xt
         return return_type(std::move(buf),
                            xtl::forward_sequence<typename return_type::inner_shape_type, SC>(shape),
                            xtl::forward_sequence<typename return_type::inner_strides_type, SS>(strides));
+    }
+
+    /**
+     * Contructs an xarray_adaptor of the given C array allocated on the stack, with the
+     * specified shape and layout.
+     * @param c_array the C array allocated on the stack
+     * @param shape the shape of the xarray_adaptor
+     * @param l the layout_type of the xarray_adaptor
+     */
+    template <layout_type L = XTENSOR_DEFAULT_LAYOUT, class T, std::size_t N, class SC,
+              XTL_REQUIRES(detail::not_an_array<std::decay_t<SC>>)>
+    inline auto adapt(T (&c_array)[N], const SC& shape, layout_type l = L)
+    {
+        return adapt(&c_array[0], N, xt::no_ownership(), shape, l);
+    }
+
+    /**
+     * Contructs an xarray_adaptor of the given C array allocated on the stack, with the
+     * specified shape and stirdes.
+     * @param c_array the C array allocated on the stack
+     * @param shape the shape of the xarray_adaptor
+     * @param strides the strides of the xarray_adaptor
+     */
+    template <class T, std::size_t N, class SC, class SS,
+             XTL_REQUIRES(detail::not_an_array<std::decay_t<SC>>,
+                          detail::not_a_layout<std::decay_t<SS>>)>
+    inline auto adapt(T (&c_array)[N], SC&& shape, SS&& strides)
+    {
+        return adapt(&c_array[0], N, xt::no_ownership(),
+                     std::forward<SC>(shape),
+                     std::forward<SS>(strides));
     }
 
     /***************************
@@ -308,7 +339,37 @@ namespace xt
                            xtl::forward_sequence<typename return_type::inner_shape_type, SC>(shape),
                            xtl::forward_sequence<typename return_type::inner_strides_type, SS>(strides));
     }
+    
+    /**
+     * Contructs an xtensor_adaptor of the given C array allocated on the stack, with the
+     * specified shape and layout.
+     * @param c_array the C array allocated on the stack
+     * @param shape the shape of the xarray_adaptor
+     * @param l the layout_type of the xarray_adaptor
+     */
+    template <layout_type L = XTENSOR_DEFAULT_LAYOUT, class T, std::size_t N, class SC,
+              XTL_REQUIRES(detail::is_array<std::decay_t<SC>>)>
+    inline auto adapt(T (&c_array)[N], const SC& shape, layout_type l = L)
+    {
+        return adapt(&c_array[0], N, xt::no_ownership(), shape, l);
+    }
 
+    /**
+     * Contructs an xtensor_adaptor of the given C array allocated on the stack, with the
+     * specified shape and stirdes.
+     * @param c_array the C array allocated on the stack
+     * @param shape the shape of the xarray_adaptor
+     * @param strides the strides of the xarray_adaptor
+     */
+    template <class T, std::size_t N, class SC, class SS,
+             XTL_REQUIRES(detail::is_array<std::decay_t<SC>>,
+                          detail::not_a_layout<std::decay_t<SS>>)>
+    inline auto adapt(T (&c_array)[N], SC&& shape, SS&& strides)
+    {
+        return adapt(&c_array[0], N, xt::no_ownership(),
+                     std::forward<SC>(shape),
+                     std::forward<SS>(strides));
+    }
     /**
      * Constructs an non-owning xtensor_fixed_adaptor from a pointer with the
      * specified shape and layout.

--- a/test/test_xadapt.cpp
+++ b/test/test_xadapt.cpp
@@ -78,6 +78,31 @@ namespace xt
         EXPECT_EQ(1, data2[2]);
     }
 
+    TEST(xarray_adaptor, c_stack_array)
+    {
+        double data[4] = { 1., 2., 3., 4. };
+        using shape_type = std::vector<vec_type::size_type>;
+        shape_type s({2, 2});
+        auto a = adapt(data, s);
+
+        EXPECT_EQ(a.size(), 4u);
+        if(XTENSOR_DEFAULT_LAYOUT == xt::layout_type::row_major)
+        {
+            EXPECT_EQ(a(0, 1), 2.);
+            EXPECT_EQ(a(1, 0), 3.);
+        }
+        else
+        {
+            EXPECT_EQ(a(0, 1), 3.);
+            EXPECT_EQ(a(1, 0), 2.);
+        }
+        shape_type str({2, 1});
+        auto b = adapt(data, s, str);
+        EXPECT_EQ(b.size(), 4u);
+        EXPECT_EQ(b(0, 1), 2.);
+        EXPECT_EQ(b(1, 0), 3.);
+    }
+
     TEST(xarray_adaptor, no_ownership_assign)
     {
         size_t size = 1;
@@ -247,6 +272,32 @@ namespace xt
         auto a2 = adapt(std::move(data2), size, acquire_ownership(), s, str);
         a2(1, 0) = 1;
         EXPECT_EQ(1, data2[2]);
+    }
+
+    TEST(xtensor_adaptor, c_stack_array)
+    {
+        double data[4] = { 1., 2., 3., 4. };
+        using shape_type = std::array<vec_type::size_type, 2>;
+        shape_type s = {2, 2};
+        auto a = adapt(data, s);
+
+        EXPECT_EQ(a.size(), 4u);
+        if(XTENSOR_DEFAULT_LAYOUT == xt::layout_type::row_major)
+        {
+            EXPECT_EQ(a(0, 1), 2.);
+            EXPECT_EQ(a(1, 0), 3.);
+        }
+        else
+        {
+            EXPECT_EQ(a(0, 1), 3.);
+            EXPECT_EQ(a(1, 0), 2.);
+        }
+
+        shape_type str = {2, 1};
+        auto b = adapt(data, s, str);
+        EXPECT_EQ(b.size(), 4u);
+        EXPECT_EQ(b(0, 1), 2.);
+        EXPECT_EQ(b(1, 0), 3.);
     }
 
     TEST(xtensor_adaptor, no_ownership_assign)

